src/Raven.Server/Documents/Patch/PatchDocumentCommand.cs

```diff
@@ -461,24 +467,43 @@ public class PatchDocumentCommand : PatchDocumentCommandBase
             _ignoreMaxStepsForScript = ignoreMaxStepsForScript;
             if (string.IsNullOrEmpty(id) || id.EndsWith(identityPartsSeparator) || id.EndsWith('|'))
                 throw new ArgumentException($"The ID argument has invalid value: '{id}'", nameof(id));
+
+            // If the caller is a DocumentsOperationContext, then we can apply the optimization.
+            if (context is DocumentsOperationContext doContext)
+            {
+                _database = doContext.DocumentDatabase;
+                _returnRun = _database.Scripts.GetScriptRunner(_patch.Run, readOnly: false, out _run);
+                _disposableStatement = _ignoreMaxStepsForScript ? _run.ScriptEngine.DisableMaxStatements() : null;
+                _disposableScriptRunner = _patchIfMissing.Run != null ? _database.Scripts.GetScriptRunner(_patchIfMissing.Run, readOnly: false, out _runIfMissing) : null;
+            }
         }
 
         protected override long ExecuteCmd(DocumentsOperationContext context)
         {
-            ScriptRunner.SingleRun runIfMissing = null;
-            _database = context.DocumentDatabase;
-
-            using (_database.Scripts.GetScriptRunner(_patch.Run, readOnly: false, out var run))
+            if (_database == null)
```

Comment:
Note that you are checking on _different_ conditional than the ctor, that can lead to bugs down the line

Reply:
It may happen that constructor would not receive that context and therefore doesnt have an actual instance, so we are doing the very same thing here. Do you mean that we should refactor this to use the same 'method' for both?

Reply:
You are checking `_context` in the ctor, and `_database` here.

Reply:
Yes, I rewrote that to make it easier to follow. The initialization what it does is store the `_database` instance, so we are checking that to know if it has been initialized before or not.